File: ios/LowerBackStretching/Views/StretchAnimationView.swift
import SwiftUI

/// The player surface that replaces the YouTube embed. If `animation` is
/// present we render a looping stick-figure that demonstrates the stretch;
/// otherwise we show a static placeholder. In both cases, if `youtubeId`
/// points to a real video we surface a small "Watch demo on YouTube" link
/// that opens the video in the YouTube app or browser — embedded playback
/// was removed because YouTube's mid-2025 embedder-verification tightening
/// (Error 152) made WebView-based embeds unreliable across devices.
struct StretchAnimationView: View {
    let animation: StretchAnimationSpec?
    let youtubeId: String

    var body: some View {
        ZStack {
            RoundedRectangle(cornerRadius: 12)
                .fill(Color(.secondarySystemBackground))

            if let animation, !animation.poses.isEmpty {
                AnimatedFigure(spec: animation)
                if Self.hasWatchableVideo(youtubeId) {
                    VStack {
                        Spacer()
                        HStack {
                            Spacer()
                            WatchOnYouTubeChip(youtubeId: youtubeId)
                                .padding(8)
                        }
                    }
                }
            } else {
                NoAnimationPlaceholder(youtubeId: youtubeId)
            }
        }
        .clipShape(RoundedRectangle(cornerRadius: 12))
    }

    static func hasWatchableVideo(_ id: String) -> Bool {
        !id.isEmpty && !id.hasPrefix("PLACEHOLDER_")
    }
}

private struct AnimatedFigure: View {
    let spec: StretchAnimationSpec

    var body: some View {
        TimelineView(.animation) { context in
            let t = Self.tValue(
                date: context.date,
                loopSeconds: max(spec.loopSeconds, 0.25),
                poseCount: spec.poses.count
            )
            let joints = StretchAnimationView.interpolatedPose(poses: spec.poses, t: t)
            Canvas { ctx, size in
                StretchAnimationView.draw(joints: joints, in: ctx, size: size)
            }
        }
    }

    /// Maps wall-clock time onto `[0, poseCount)`, advancing through the
    /// loop at `loopSeconds` per full cycle.
    static func tValue(date: Date, loopSeconds: Double, poseCount: Int) -> Double {
        let secs = date.timeIntervalSinceReferenceDate
        let phase = secs.truncatingRemainder(dividingBy: loopSeconds) / loopSeconds
        return phase * Double(poseCount)
    }
}

extension StretchAnimationView {
    /// Bones connect consecutive joint pairs as line segments. The
    /// L-suffixed joints are optional — only used by front-view poses
    /// that need to show both sides of the body. Side-view poses omit
    /// them and the renderer's guard else continue skips.
    static let skeletonBones: [(String, String)] = [
        ("head", "neck"),
        ("neck", "spineMid"),
        ("spineMid", "hip"),
        ("neck", "shoulder"),
        ("shoulder", "elbow"),
        ("elbow", "hand"),
        ("hip", "knee"),
        ("knee", "foot"),
        // Optional left side for front-view poses.
        ("neck", "shoulderL"),
        ("shoulderL", "elbowL"),
        ("elbowL", "handL"),
        ("hip", "kneeL"),
        ("kneeL", "footL"),
    ]

    /// Interpolates between consecutive poses with an eased segment,
    /// wrapping after the last pose. `t` is expected in `[0, poses.count)`.
    static func interpolatedPose(poses: [Pose], t: Double) -> [String: CGPoint] {
        guard !poses.isEmpty else { return [:] }
        let n = poses.count
        let wrapped = (t.truncatingRemainder(dividingBy: Double(n)) + Double(n))
            .truncatingRemainder(dividingBy: Double(n))
        let i = min(max(Int(wrapped), 0), n - 1)
        let j = (i + 1) % n
        let raw = wrapped - Double(i)
        let eased = max(0.0, min(1.0, 0.5 - 0.5 * cos(raw * .pi)))
        let a = poses[i].joints
        let b = poses[j].joints
        var out: [String: CGPoint] = [:]
        for key in a.keys {
            guard let pa = a[key], pa.count >= 2 else { continue }
            let pb = b[key] ?? pa
            guard pb.count >= 2 else { continue }
            out[key] = CGPoint(
                x: lerp(pa[0], pb[0], eased),
                y: lerp(pa[1], pb[1], eased)
            )
        }
        return out
    }

    static func draw(joints: [String: CGPoint], in ctx: GraphicsContext, size: CGSize) {
        let w = size.width
        let h = size.height
        let minDim = min(w, h)
        let stroke = max(3.0, minDim * 0.018)
        let headRadius = minDim * 0.05
        let color = Color.accentColor

        let absolute: [String: CGPoint] = joints.mapValues { p in
            CGPoint(x: p.x * w, y: p.y * h)
        }

        var path = Path()
        for (aName, bName) in skeletonBones {
            guard let a = absolute[aName], let b = absolute[bName] else { continue }
            path.move(to: a)
            path.addLine(to: b)
        }
        ctx.stroke(
            path,
            with: .color(color),
            style: StrokeStyle(lineWidth: stroke, lineCap: .round)
        )

        if let head = absolute["head"] {
            let rect = CGRect(
                x: head.x - headRadius,
                y: head.y - headRadius,
                width: headRadius * 2,
                height: headRadius * 2
            )
            ctx.stroke(
                Path(ellipseIn: rect),
                with: .color(color),
                lineWidth: stroke
            )
        }
    }
}

private func lerp(_ a: Double, _ b: Double, _ t: Double) -> Double {
    a + (b - a) * t
}

private struct NoAnimationPlaceholder: View {
    let youtubeId: String

    var body: some View {
        VStack(spacing: 8) {
            Image(systemName: "figure.cooldown")
                .font(.system(size: 36))
                .foregroundStyle(.secondary)
            Text("Follow the description and timer below.")
                .font(.body)
                .foregroundStyle(.secondary)
                .multilineTextAlignment(.center)
            if StretchAnimationView.hasWatchableVideo(youtubeId) {
                WatchOnYouTubeChip(youtubeId: youtubeId)
                    .padding(.top, 4)
            }
        }
        .padding(16)
    }
}

private struct WatchOnYouTubeChip: View {
    let youtubeId: String

    @Environment(\.openURL) private var openURL

    var body: some View {
        Button {
            if let url = URL(string: "https://www.youtube.com/watch?v=\(youtubeId)") {
                openURL(url)
            }
        } label: {
            Label("Watch demo", systemImage: "play.circle")
                .font(.caption.weight(.medium))
                .padding(.horizontal, 10)
                .padding(.vertical, 6)
        }
        .buttonStyle(.bordered)
        .controlSize(.small)
        .tint(.accentColor)
    }
}

File: android/app/src/main/java/com/lowerbackstretching/ui/components/StretchAnimationView.kt
package com.lowerbackstretching.ui.components

import android.content.Intent
import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.PlayCircleOutline
import androidx.compose.material.icons.filled.SelfImprovement
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.core.net.toUri
import com.lowerbackstretching.core.SKELETON_BONES
import com.lowerbackstretching.core.interpolatedPose
import com.lowerbackstretching.core.model.StretchAnimationSpec

/**
 * The player surface that replaces the YouTube embed. If [animation] is
 * present we render a looping stick-figure that demonstrates the stretch;
 * otherwise we show a static placeholder. In both cases, if [youtubeId]
 * points to a real video we surface a small "Watch demo on YouTube" link
 * that opens the video in the YouTube app or browser — embedded playback
 * was removed because YouTube's mid-2025 embedder-verification tightening
 * (Error 152) made WebView-based embeds unreliable across devices.
 */
@Composable
fun StretchAnimationView(
    animation: StretchAnimationSpec?,
    youtubeId: String,
    videoUrl: String? = null,
    modifier: Modifier = Modifier,
) {
    val surface = MaterialTheme.colorScheme.surfaceVariant
    Box(
        modifier = modifier
            .aspectRatio(16f / 9f)
            .clip(RoundedCornerShape(12.dp))
            .background(surface),
    ) {
        if (!videoUrl.isNullOrBlank()) {
            StretchVideoPlayer(
                videoUrl = videoUrl,
                modifier = Modifier.fillMaxSize(),
            )
        } else if (animation != null && animation.poses.isNotEmpty()) {
            AnimatedFigure(
                spec = animation,
                color = MaterialTheme.colorScheme.primary,
                modifier = Modifier.fillMaxSize(),
            )
            if (hasWatchableVideo(youtubeId)) {
                WatchOnYouTubeChip(
                    youtubeId = youtubeId,
                    modifier = Modifier
                        .align(Alignment.BottomEnd)
                        .padding(8.dp),
                )
            }
        } else {
            NoAnimationPlaceholder(youtubeId = youtubeId)
        }
    }
}

@Composable
private fun AnimatedFigure(
    spec: StretchAnimationSpec,
    color: Color,
    modifier: Modifier = Modifier,
) {
    val transition = rememberInfiniteTransition(label = "stretch")
    val durationMs = (spec.loopSeconds * 1000).toInt().coerceAtLeast(250)
    val t by transition.animateFloat(
        initialValue = 0f,
        targetValue = spec.poses.size.toFloat(),
        animationSpec = infiniteRepeatable(
            animation = tween(durationMillis = durationMs, easing = LinearEasing),
            repeatMode = RepeatMode.Restart,
        ),
        label = "poseT",
    )
    val pose = interpolatedPose(spec.poses, t)
    Canvas(modifier = modifier) {
        drawFigure(pose, color)
    }
}

private fun DrawScope.drawFigure(
    joints: Map<String, Pair<Float, Float>>,
    color: Color,
) {
    val w = size.width
    val h = size.height
    val strokeWidth = (size.minDimension * 0.018f).coerceAtLeast(3f)
    val headRadius = size.minDimension * 0.05f
    val absolute = joints.mapValues { (_, p) -> Offset(p.first * w, p.second * h) }

    for ((aName, bName) in SKELETON_BONES) {
        val a = absolute[aName] ?: continue
        val b = absolute[bName] ?: continue
        drawLine(
            color = color,
            start = a,
            end = b,
            strokeWidth = strokeWidth,
            cap = StrokeCap.Round,
        )
    }
    absolute["head"]?.let { c ->
        drawCircle(color = color, radius = headRadius, center = c, style = Stroke(width = strokeWidth))
    }
}

@Composable
private fun NoAnimationPlaceholder(youtubeId: String) {
    Column(
        modifier = Modifier
            .fillMaxSize()
            .padding(16.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center,
    ) {
        Icon(
            imageVector = Icons.Filled.SelfImprovement,
            contentDescription = null,
            tint = MaterialTheme.colorScheme.onSurfaceVariant,
            modifier = Modifier.size(48.dp),
        )
        Spacer(Modifier.height(8.dp))
        Text(
            text = "Follow the description and timer below.",
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            textAlign = TextAlign.Center,
        )
        if (hasWatchableVideo(youtubeId)) {
            Spacer(Modifier.height(8.dp))
            WatchOnYouTubeChip(youtubeId = youtubeId)
        }
    }
}

@Composable
private fun WatchOnYouTubeChip(
    youtubeId: String,
    modifier: Modifier = Modifier,
) {
    val context = LocalContext.current
    Surface(
        modifier = modifier,
        color = MaterialTheme.colorScheme.surface.copy(alpha = 0.85f),
        shape = RoundedCornerShape(16.dp),
        tonalElevation = 2.dp,
    ) {
        TextButton(
            onClick = {
                val intent = Intent(
                    Intent.ACTION_VIEW,
                    "https://www.youtube.com/watch?v=$youtubeId".toUri(),
                ).addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                context.startActivity(intent)
            },
            modifier = Modifier.padding(horizontal = 4.dp),
        ) {
            Icon(
                imageVector = Icons.Filled.PlayCircleOutline,
                contentDescription = null,
                modifier = Modifier.size(18.dp),
            )
            Spacer(Modifier.width(6.dp))
            Text(
                text = "Watch demo",
                style = MaterialTheme.typography.labelMedium,
            )
        }
    }
}

private fun hasWatchableVideo(youtubeId: String): Boolean =
    youtubeId.isNotBlank() && !youtubeId.startsWith("PLACEHOLDER_")
